src/hpr_packet_reporter.erl

```diff
@@ -0,0 +1,331 @@
+    restart_report_interval/1,
+    stop_report_interval/0
+]).
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2,
+    code_change/3
+]).
+
+-define(SERVER, ?MODULE).
+-define(ETS, hpr_packet_report_ets).
+
+%% 5 minute interval
+-define(DEFAULT_REPORT_INTERVAL, 300000).
+-define(MAX_FILE_SIZE, 50_000_000).
+
+-type interval_duration_ms() :: non_neg_integer().
+
+-record(state, {
+    aws_client :: aws_client:aws_client(),
+    file_path :: string(),
+    report_interval :: timer:tref() | undefined
+}).
+
+%%%===================================================================
+%%% API Functions
+%%%===================================================================
+
+start_link(Args) ->
+    gen_server:start_link({local, ?SERVER}, ?SERVER, Args, []).
+
+-spec init_ets() -> ok.
+init_ets() ->
+    ?ETS = ets:new(?ETS, [named_table, bag, public, {write_concurrency, true}]),
+    ok.
+
+-spec report_packet(hpr_packet_up:packet(), hpr_route:route()) -> ok.
+report_packet(Packet, Route) ->
+    EncodedPacket = encode_packet(Packet, Route),
+    Timestamp = erlang:system_time(millisecond),
+    true = ets:insert(?ETS, {Timestamp, EncodedPacket}),
+    ok.
+
+-spec write_packets() -> ok.
+write_packets() ->
+    gen_server:cast(?SERVER, write).
+
+-spec upload_packets(string()) -> ok.
+upload_packets(FilePath) ->
+    gen_server:cast(?SERVER, {upload, FilePath}).
+
+-spec restart_report_interval(interval_duration_ms()) -> ok.
+restart_report_interval(IntervalDuration) ->
+    gen_server:cast(?SERVER, {start_interval, IntervalDuration}).
+
+stop_report_interval() ->
+    gen_server:cast(?SERVER, stop_interval).
+
+%%%===================================================================
+%%% Spawning and gen_server implementation
+%%%===================================================================
+
+init(#{base_dir := BaseDir} = _Args) ->
+    AWSClient = setup_aws(),
+    WriteDir = filename:join(BaseDir, "tmp/"),
+    ok = filelib:ensure_dir(WriteDir),
+    %% TODO: File rotation
+    TempFilePath = filename:join(WriteDir, "packetreport.gz"),
+    {ok, ReportInterval} = start_report_interval(),
+    {ok, #state{
+        aws_client = AWSClient,
+        file_path = TempFilePath,
+        report_interval = ReportInterval
+    }}.
+
+handle_call(Request, From, State = #state{}) ->
+    lager:warning("rcvd unknown call msg: ~p from: ~p", [Request, From]),
+    {reply, ok, State}.
+
+handle_cast(write, State = #state{file_path = FilePath}) ->
+    {ok, FileSize, MaxFileSize} = handle_write(FilePath),
+
+    case FileSize >= MaxFileSize of
+        true -> ?MODULE:upload_packets(FilePath);
+        false -> skip
+    end,
+    {noreply, State};
+handle_cast({upload, FilePath}, State = #state{aws_client = AWSClient}) ->
+    Timestamp = erlang:system_time(millisecond),
+    FileName = list_to_binary("packetreport." ++ integer_to_list(Timestamp) ++ ".gz"),
+    upload_file(AWSClient, FilePath, FileName),
```

Comment:
We shouldn't delete the file if we can't upload it. 
Maybe a TODO here for some retry logic similar to what they're doing in rust.

Reply:
👍🏻